### PR TITLE
Explore: Align multiple log volumes

### LINKED ISSUE #65492: Logs: Improve UX of multiple log volumes (mixed data source)
### Background

#61546 introduced basic support for log volume in mixed data source. The feature is usable, thought we identified a few UX tweaks that can improve the experience.

### Requirements ###

High level: _Align graphs on x and y axes_

#### Technical details

Each graph should have the same maximum on Y axis to ensure all bars have the same size relative to the amount of log lines they represent.

Each graph should span the same time range so the crosshair cursor is aligned when hovering over the data.

### PR DESCRIPTION
**Pull request by ifrost**
This PR aligns log volumes histogram to use the same values on Y and X axes:

| before | after | 
|----|----|
| Screenshot 2023-03-07 at 22 00 10 | Screenshot 2023-03-07 at 21 58 26 |

Note: Helper functions for log volume were moved from `grafana-data/types` to `features/logs/utils`. These were recently added in https://github.com/grafana/grafana/pull/63036, not released, hence they can be ignored as breaking changes.

Fixes #65492

### CODE REVIEW

**Comment by ifrost** [comment]
> Possible breaking changes

Helper functions for log volume were moved from `grafana-data/types` to `features/logs/utils`. These were recently added in https://github.com/grafana/grafana/pull/63036, not released, hence they can be ignored as breaking changes.

**Review by ivanahuckova — CHANGES_REQUESTED** [review]
Left some comments bellow. I also discovered new issue this PR creates - the zooming in logs volume does not work anymore. 

https://user-images.githubusercontent.com/30407135/226338635-e05e4ade-b20e-43c9-b01c-912c554ebed1.mov

File: public/app/features/explore/Graph/ExploreGraph.tsx [hunk 1]
```diff
@@ -54,6 +54,7 @@ interface Props {
   onChangeTime: (timeRange: AbsoluteTimeRange) => void;
   graphStyle: ExploreGraphStyle;
   anchorToZero?: boolean;
+  maximum?: number;
```
[the remark this answers is not in the document]
Comment (Elfo404) [reply public/app/features/explore/Graph/ExploreGraph.tsx:57]: +1

**Review by Elfo404 — APPROVED** [review]
seems to work as expected, i agree with ivana that naming may be a bit better, so if you could check that it'd be great, otherwise lgtm

File: public/app/features/explore/Graph/ExploreGraph.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/features/explore/Graph/ExploreGraph.tsx]: Updated in f41a5f00d652737b20de5dc9e21b1e6fdafbc50a

File: public/app/features/logs/utils.ts [hunk 2]
```diff
@@ -149,3 +160,51 @@ export function logRowsToReadableJson(logs: LogRowModel[]) {
...
+export const getLogsVolumeDimensions = (
+  dataFrames: DataFrame[],
+  defaultRange: AbsoluteTimeRange
+): { maximum: number; range: AbsoluteTimeRange } => {
+  let widestRange: AbsoluteTimeRange | undefined;
+  let maximumValue = -Infinity;
+
+  dataFrames.forEach((dataFrame: DataFrame) => {
+    const dataFrameRange = dataFrame.meta?.custom?.absoluteRange || defaultRange;
+    if (dataFrameRange) {
```
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/features/logs/utils.ts]: This was all wrong 😱  Updated in c0fc6f10ee1d4eebda917a84977fdf4a10101bf7

**Review by gelicia — APPROVED** [review]
I think you need to add `reuseWhenZoomingIn` here https://github.com/grafana/grafana/blob/main/public/app/features/explore/utils/supplementaryQueries.ts#L94 

but other than that, it looks good! Great work!

Tested with gdev-loki and test data on the logs scenario. Time range "last 5 minutes" shows logs volume data for both, with the same x and y scales.

**Comment by ifrost** [comment]
There's also a change in the behavior of limited log volume. With this change in place, it will always be displayed with the selected range, not narrowed down to the results even if there's only one limited log volume visible (with and without mixed DS). 

@grafana/observability-logs I feel this actually makes everything more predictable, but do you think it's okay to make this change? Or should we have a special check - if there's only a limited log volume => change the visible range of the log volume to start with the first logline?

**Comment by ifrost** [comment]
Converting back to draft as we want https://github.com/grafana/grafana/pull/65392 to be merged first.

**Comment by ifrost** [comment]
Reopening for review (since https://github.com/grafana/grafana/pull/65392 got merged).

cc @grafana/observability-logs

**Review by ivanahuckova — COMMENTED** [review]
1 small nit UX issue I noticed - when you zoom in and `FullRange` logs volume does not receive data, the order of panels switched. Is there a way we can guarantee order of the volumes when zooming in?
 
https://user-images.githubusercontent.com/30407135/230400817-9a3c64b5-cf60-4aef-ad8f-aef20cfe5ba2.mov

**Comment by ifrost** [comment]
> 1 small nit UX issue I noticed - when you zoom in and `FullRange` logs volume does not receive data, the order of panels switched. Is there a way we can guarantee order of the volumes when zooming in?

Added a50a9d984217374eae91a90815045690da836654 to sort volumes by data source name

File: public/app/features/explore/utils/supplementaryQueries.ts [hunk 3]
```diff
@@ -95,6 +95,9 @@ const createFallbackLogVolumeProvider = (
               absoluteRange: exploreData.logsResult?.visibleRange!,
               datasourceName,
               sourceQuery: queryTargets.find((query) => query.refId === refId)!,
+              // If log volume is based on returned log lines, zooming in may return different
+              // results, so we don't want to reuse the data
+              reuseWhenZoomingIn: false,
```
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/features/explore/utils/supplementaryQueries.ts]: Good idea! Changed in e8abd9495dc11fa02890cccf0b453cd3300cf93b

**Review by ivanahuckova — APPROVED** [review]
LGTM!

**Pull request merged**